### ISSUE TITLE
"Neural Networks: Zero to Hero" by A. Karpathy

### ISSUE THREAD

**Issue by aigoncharov** [aigoncharov/ml_misfits#6]
**About:** Following the legendary series ["Neural Networks: Zero to Hero"](https://www.youtube.com/playlist?list=PLAqhIrjkxbuWI23v9cThsA9GvCAUhRvKZ) by A. Karpathy
**Schedule:** TBD

**What this thread is for:**
- Announcing homework tasks and theoretical content;
- Discussing questions related to the course;
- Posting links to complete homeworks for review.

_Please, refrain from spamming! It forces others to turn off notifications and, subsequently, leads to a slow and painful death of the community!_

**Comment by aigoncharov** [comment]
# HW1

**Deadline**: July 8
**Tasks**:
- Watch https://youtu.be/VMj-3S1tku0?si=KWRNO_pkYyKe3P-X
- Complete[ this Google colab](https://colab.research.google.com/drive/1FPTx1RXtBfc4MaTkf7viZZD4U2F9gtKN?usp=sharing)

Add 👍 if you are going to join!

**Comment by aigoncharov** [comment]
Due to a popular demand, meetings have been move to Friday. See you on July 11!

**Comment by aigoncharov** [comment]
HW1 https://github.com/aigoncharov/ml_misfits/pull/11
Review and comments are welcome!

**Comment by aigoncharov** [comment]
# HW2

**Deadline**: July 11
**Lecture**:
- Watch https://youtu.be/PaCmpygFfXo?si=heMXDWRA3YUz3Qsi
**HW:**
E01: train a trigram language model, i.e. take two characters as an input to predict the 3rd one. Feel free to use either counting or a neural net. Evaluate the loss; Did it improve over a bigram model?
E02: split up the dataset randomly into 80% train set, 10% dev set, 10% test set. Train the bigram and trigram models only on the training set. Evaluate them on dev and test splits. What can you see?
E03: use the dev set to tune the strength of smoothing (or regularization) for the trigram model - i.e. try many possibilities and see which one works best based on the dev set loss. What patterns can you see in the train and dev set loss as you tune this strength? Take the best setting of the smoothing and evaluate on the test set once and at the end. How good of a loss do you achieve?
E04: we saw that our 1-hot vectors merely select a row of W, so producing these vectors explicitly feels wasteful. Can you delete our use of F.one_hot in favor of simply indexing into rows of W?
E05: look up and use F.cross_entropy instead. You should achieve the same result. Can you think of why we'd prefer to use F.cross_entropy instead?

Add 👍 if you are going to join!

**Comment by aigoncharov** [comment]
Moving this meeting to async scheduling without a regular call due to low popularity. Still going to post my HW solutions as I go through them. Questions are welcome!
